README.md
# MMM-RBB-Weather

[![Build Status](https://travis-ci.com/nkl-kst/MMM-RBB-Weather.svg?branch=master)](https://travis-ci.com/nkl-kst/MMM-RBB-Weather)
[![Coverage Status](https://coveralls.io/repos/github/nkl-kst/MMM-RBB-Weather/badge.svg)](https://coveralls.io/github/nkl-kst/MMM-RBB-Weather)
[![Maintainability](https://api.codeclimate.com/v1/badges/a8567d7917169f2cb1be/maintainability)](https://codeclimate.com/github/nkl-kst/MMM-RBB-Weather/maintainability)
[![Greenkeeper](https://badges.greenkeeper.io/nkl-kst/MMM-RBB-Weather.svg)](https://greenkeeper.io/)
[![dependencies Status](https://david-dm.org/nkl-kst/MMM-RBB-Weather/status.svg)](https://david-dm.org/nkl-kst/MMM-RBB-Weather)
[![devDependencies Status](https://david-dm.org/nkl-kst/MMM-RBB-Weather/dev-status.svg)](https://david-dm.org/nkl-kst/MMM-RBB-Weather?type=dev)
[![Known Vulnerabilities](https://snyk.io/test/github/nkl-kst/MMM-RBB-Weather/badge.svg?targetFile=package.json)](https://snyk.io/test/github/nkl-kst/MMM-RBB-Weather?targetFile=package.json)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](http://choosealicense.com/licenses/mit)

This is a module for the awesome [MagicMirror²](https://github.com/MichMich/MagicMirror/). It displays local weather data for cities in Berlin and Brandenburg (Germany) based on [RBB weather data](https://www.rbb24.de/wetter/wetterkarte/).

![Screenshot](.github/screen.gif)

## Install the module

Go to modules folder
```
cd modules/
```

Clone this module from Github
```
git clone https://github.com/nkl-kst/MMM-RBB-Weather
```

Switch to newly created module folder
```
cd MMM-RBB-Weather/
```

Install dependencies
```
npm install --only=prod
```

After adding this module to your config (see below) restart your MagicMirror.

## Update the module

Go to modules folder

```
cd modules/MMM-RBB-Weather
```

Pull changes from Github

```
git pull
```

Install new dependencies

```
npm install --only=prod
```

Since this repository ignores the automatically generated ``package-lock.json``, pulling changes should always work. If not, try to reset your module with ``git reset --hard`` before pulling new changes.

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-RBB-Weather',
            position: "top_right", // All available positions
            config: {
                // See below for configurable options, this is optional
            }
        }
    ]
}
```

## Configuration options

All options are optional so the module works out of the box.

| Option                 | Description
|----------------------- |-----------
| `id`                   | City ID for data, see [table below](#city-ids)<br><br>**Type:** `String`<br>**Default:** `10381` (Berlin-Steglitz)
| `days`                 | Days shown in forecast table. Set this to `0` to display only current weather. Data for seven days are available (including today), so `7` is the maximum here. <br><br>**Type:** `Number`<br>**Default:** `4`
| `animationSpeed`       | Duration of content refresh animation in seconds.<br><br>**Type:** `Number`<br>**Default:** `1`
| `updateInterval`       | Time between loading new weather data in seconds.<br><br>**Type:** `Number`<br>**Default:** `600` (10 minutes)
| `showCurrentText`      | Flag to display current weather text.<br><br>**Type:** `Boolean`<br>**Default:** `true`
| `showCurrentWindspeed` | Flag to display current windspeed information.<br><br>**Type:** `Boolean`<br>**Default:** `true`
| `showRainProbability`  | Flag to display rain probability in forecast table.<br><br>**Type:** `Boolean`<br>**Default:** `true`
| `showUpdateTime`       | Flag to display last data update time.<br><br>**Type:** `Boolean`<br>**Default:** `false`
| `showWindspeed`        | Flag to display windspeed in forecast table.<br><br>**Type:** `Boolean`<br>**Default:** `false`
| `animateCurrentIcon`   | Flag to animate icon for current data.<br><br>**Type:** `Boolean`<br>**Default:** `true`
| `animateForecastIcon`  | Flag to animate icons in forecast table.<br><br>**Type:** `Boolean`<br>**Default:** `false`
| `dayFormat`            | Day format in forecast table, see [Moment.js formats](https://momentjs.com/docs/#/displaying/format/) for details.<br><br>**Type:** `String`<br>**Default:** `ddd`
| `splitCurrentTextGreater` | Split current weather text if it is larger than this value. Set it to `0` to disable splitting.<br><br>**Type:** `Number`<br>**Default:** `30`
| `tableClass`           | Classes added to forecast table. Could be used for additional styling, sizing etc.<br><br>**Type:** `String`<br>**Default:** `small`
| `whiteIcons`           | Flag to convert weather icons to simple white icons.<br><br>**Type:** `Boolean`<br>**Default:** `true`
| `triggers`             | Array of triggers to show/hide other modules based on weather data.<br><br>**Type:** `Array` of trigger objects<br>**Default:** `[]`

### Triggers

You can define triggers to show or hide other modules based on weather data. A trigger object is defined by the following keys:

| Key                    | Description
|----------------------- |-----------
| `day`                  | Day to check `field` for fulfilled trigger condition. Day `0` holds current values, day `1` represents today, day `2` is tomorrow and so on until day `7`.<br><br>**Type:** `Number`
| `field`                | Field to check `value` for fulfilled trigger condition. Possible fields are:<br><br><ul><li>`temp` - Temperature (only for day `0`)</li><li>`maxtemp` - Maximum temperature (only for days `1` to `7`)</li><li>`mintemp` - Minimum temperature (only for days `1` to `7`)</li><li>`ffkmh` - Windspeed</li><li>`prr` - Rain possibility (only for days `1` to `7`)</li></ul><br>**Type:** `String`
| `value`                | Show module if field data is higher than this value.<br><br>**Type:** `Number`<br>
| `module`               | Name of the module you want to show/hide. Modules are determined by their class name, so you could define a class in module configs to trigger more than one module.<br><br>**Type:** `String`<br>
| `hide`                 | Use this option to hide a module instead of showing it if trigger conditions are fulfilled.<br><br>**Type:** `Boolean`<br>**Default:** `false`

This example trigger hides the clock module if todays rain possiblity is higher than 50 percent, otherwise the clock module is shown:

```
triggers: [
    { day: 1, field: 'prr', value: 50, module: 'clock', hide: true }
]
```

## Known issues

### MagicMirror crashs after several hours
The animated icons consume a high amount of memory, resulting in an Electron crash and therefore a black MagicMirror screen. If you experience this issue, please use static icons instead (as described in [Configuration options](#configuration-options)). Refer to [issue #16](https://github.com/nkl-kst/MMM-RBB-Weather/issues/16) for more information.

### Suddenly no data
There were some changes in the weather data in October 2020, primarly city IDs have changed. Please refer to the [table below](#city-ids) and get your city ID.

## Problems

If you have any problems or questions, feel free to open an issue. There are many possible improvements for this module so please let me know if you miss something.

## Developer notes

To run all unit tests just fire this command in the module folder

```
# Install also dev dependencies
npm install

# Run tests
npm test
```

You can also check if RBB endpoints are available and still providing correct data

```
npm run test-int
```

## City IDs

Use one of these city IDs in your config.

| City                      | ID
|---------------------------|----------
| *Berlin*                  | <span/>
| Adlershof                 | `6510388`
| Marzahn                   | `3010387`
| Steglitz                  | `10381`
| Tempelhof                 | `10384`
| *Brandenburg*             | <span/>
| Altdöbern                 | `6510496`
| Angermünde                | `10291`
| Bad Liebenwerda           | `6510482`
| Baruth/Mark               | `10376`
| Bernau                    | `3210383`
| Bestensee                 | `6510375`
| Brandenburg an der Havel  | `6510371`
| Coschen                   | `3010496`
| Cottbus                   | `10496a`
| Dahme/Mark                | `6510477`
| Doberlug/Kirchhain        | `10490`
| Eberswalde                | `6510290`
| Eisenhüttenstadt          | `3010398`
| Forst (Lausitz)           | `3210497`
| Frankfurt (Oder)          | `6510399`
| Fürstenberg/Havel         | `10277`
| Fürstenwalde/Spree        | `6510395`
| Gransee                   | `3010278a`
| Guben                     | `6510497`
| Guteborn                  | `6510491`
| Herzberg (Elster)         | `10476`
| Jüterbog                  | `3010476`
| Ketzin                    | `3210380`
| Klettewitz                | `3010493a`
| Kyritz                    | `10267`
| Lenzen (Elbe)             | `3010255`
| Liebenwalde               | `3110278`
| Lieberose                 | `10496b`
| Lindneberg                | `10393`
| Lübben (Spreewald)        | `3010376`
| Luckenwalde               | `6510376b`
| Ludwigsfelde              | `6510377`
| Manschnow                 | `10396a`
| Nauen                     | `6510373`
| Neuruppin                 | `10271`
| Oranienburg               | `6510374`
| Ortrand                   | `3210488`
| Perleberg                 | `6510263`
| Potsdam                   | `10379`
| Prenzlau                  | `10289`
| Pritzwalk                 | `3010266`
| Rathenow                  | `3210370`
| Schönefeld                | `10385`
| Seelow                    | `10396b`
| Senftenberg               | `3010493b`
| Stechlin                  | `3010277`
| Strausberg                | `3210393`
| Templin                   | `6510287`
| Treuenbrietzen            | `6510376a`
| Welzow                    | `6510492`
| Wiesenburg                | `10368`
| Wittstock/Dosse           | `3010273`
| Zehdenick                 | `3010278b`
| Ziesar                    | `3010366`

## License: MIT

See [LICENSE](LICENSE.txt)


CHANGELOG.md
# MMM-RBB-Weather Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

---

## [2.3.0] - 2020-10-19

### Updated
- Updated city IDs to new XML source data
- Updated all dependencies including dev

### Fixed
- Respect empty weather data fields in XML source data


## [2.2.0] - 2019-10-12

### Added
- Optionally show update time
- Define triggers to show/hide other modules based on weather data

### Updated
- eslint packages (dev-dependencies) and resulting new linting errors/warnings


## [2.1.0] - 2019-01-26

### Added
- Added option to show/hide current weather text
- Added option to set day format in forecast table
- Added option to break current weather text if it is too large

### Fixed
- Added icon mapping for RBB icon 333001 (thunder)
- Fixed current icon position when module is placed left


## [2.0.1] - 2019-01-16

### Fixed
- Use Font Awesome 5 with correct "fas" class


## [2.0.0] - 2018-12-04

### Added
- Use RBB fallback icons if no Amchart mapping was found
- e2e tests

### Changed
- **Magic Mirror 2.2.0 is now required to run this module (Nunjucks support required)**
- Reduced range of temperature icons from 8 to 7 degrees
- Moved current text and windspeed out of the temperature div
- Retry to connect to endpoint in integration tests up to two times


## [1.2.0] - 2018-11-14

### Added
- Rendering Nunjucks templates for modules HTML ouput
- lodash.clonedeep@^4.5.0 to devDependencies
- nunjucks@^3.0.1 to devDependencies
- 'no-unused-vars' warning in eslint config

### Fixed
- Use Font Awesome (and Font Awesome 5) CSS files in getStyles()
- Use Font Awesome 5 icons not before MagicMirror version 2.6.0
- Forecast table data is now right aligned

### Updated
- Test setup with ModuleTestEnv.js and NunjucksTestEnv.js
- Moved endpoint integration tests into 'endpoint' subfolder

### Removed
- getDom() and getCurrentDiv() module functions (use Nunjucks template instead)
- decache@^4.4.0 and jsdom@^13.0.0 dependencies


## [1.1.1] - 2018-10-31

### Added
- Icon mapping for thunder and additional rain types

### Updated
- eslint-plugin-node@^8.0.0
- jsdom@^13.0.0
- Use Font Awesome 5 icons if MagicMirror version is higher than 2.5.0


## [1.1.0] - 2018-10-26

- New minor version with new features (e.g. animated icons), bugfixes etc.
- For future merges this change log should be used consistently


## [1.0.0] - 2018-09-21

- First public release


test/e2e/ModuleTest.js
const assert = require('assert');
const { setConfig, startApp, stopApp, getBrowser } = require('./MagicMirrorTestEnv');

describe('Module', function() {

    // Retries and timeouts
    this.retries(2);
    this.slow(10000);
    this.timeout(30000);

    // App browser (Electron client)
    let browser;

    before(() => {

        // Set config as environment variable
        setConfig('TestConfig.js');

        // Start Magic Mirror app
        return startApp();
    });

    after(() => {
        return stopApp();
    });

    beforeEach(async () => {

        // Wait for module data
        browser = getBrowser();
        return browser.waitForExist('div.MMM-RBB-Weather div.current');
    });

    it('should display current data', async () => {

        // Act
        const html = await browser.getHTML('div.MMM-RBB-Weather div.current');

        // Assert
        assert(html.includes('<div class="current">'));
    });

    it('should display forecast data', async () => {

        // Act
        const html = await browser.getHTML('div.MMM-RBB-Weather table.weather-table');

        // Assert
        assert(html.includes('<table class="weather-table small">'));
    });
});


test/e2e/TestConfig.js
/**
 * This is a simple Magic Mirror config used in e2e tests.
 */
var config = {

    electronOptions: {
        fullscreen: false,
        webPreferences: {
            nodeIntegration: true
        }
    },

    modules: [{
        module: 'MMM-RBB-Weather',
        position: 'top_left'
    }]
};

if (typeof module !== 'undefined') {
    module.exports = config;
}


test/int/EndpointTest.js
const assert = require('assert');

const fs = require('fs');
const path = require('path');
const https = require('https');
const libxml = require('libxmljs');

describe('Endpoints', function() {

    this.slow(1000);
    this.timeout(10000);

    describe('availability', () => {

        for (let day = 0; day <= 7; day++) {
            it(`should get data for day ${day}`, (done) => {

                // Arrange
                let url = `https://www.rbb24.de/include/wetter/data/data_bb_${day}.xml`;

                // Act
                https.get(url, (response) => {

                    // Assert
                    assert.strictEqual(response.statusCode, 200);
                    done();
                });
            });
        }

        it('should get no data for day 8', (done) => {

            // Arrange
            let url = `https://www.rbb24.de/include/wetter/data/data_bb_8.xml`;

            // Act
            https.get(url, (response) => {

                // Assert
                assert.strictEqual(response.statusCode, 404);
                done();
            });
        });
    });

    describe('content', () => {

        it('should get valid current data', (done) => {

            // Arrange
            let xsd = fs.readFileSync(path.join(__dirname, 'endpoint_current.xsd'));
            let xsdDoc = libxml.parseXml(xsd);

            let url = `https://www.rbb24.de/include/wetter/data/data_bb_0.xml`;

            // Act
            https.get(url, (response) => {

                let xml = '';
                response.on('data', (chunk) => {
                    xml += chunk;
                });

                response.on('end', () => {

                    // Parse XML
                    let xmlDoc = libxml.parseXml(xml);

                    // Assert
                    assert.strictEqual(xmlDoc.validate(xsdDoc), true);
                    done();
                });
            });
        });

        for (let day = 1; day <= 7; day++) {
            it(`should get valid forecast data for day ${day}`, (done) => {

                // Arrange
                let xsd = fs.readFileSync(path.join(__dirname, 'endpoint_forecast.xsd'));
                let xsdDoc = libxml.parseXml(xsd);

                let url = `https://www.rbb24.de/include/wetter/data/data_bb_${day}.xml`;

                // Act
                https.get(url, (response) => {

                    let xml = '';
                    response.on('data', (chunk) => {
                        xml += chunk;
                    });

                    response.on('end', () => {

                        // Parse XML
                        let xmlDoc = libxml.parseXml(xml);

                        // Assert
                        assert.strictEqual(xmlDoc.validate(xsdDoc), true);
                        done();
                    });
                });
            });
        }
    });
});


IconMapper.js
/* Magic Mirror
 * Module: MMM-RBB-Weather
 *
 * By Nikolai Keist (github.com/nkl-kst)
 * MIT Licensed.
 */

/**
 * Map RBB to Amchart icons.
 */
IconMapper = {

    /**
     * This object is used to map RBB icon names to Amchart SVG filenames. The RBB icon names are
     * set in the XML weather data returned by the RBB webservice. The Amchart files are placed in
     * vendor/amcharts folder.
     */
    iconMapping: {
        100000: 'day',
        110000: 'cloudy-day-1',
        110100: 'snowy-1',
        110300: 'snowy-1',
        111000: 'rainy-2',
        111100: 'snowy-2',
        113000: 'rainy-2',
        113300: 'snowy-2',
        120000: 'cloudy-day-1',
        120100: 'snowy-2',
        120200: 'snowy-3',
        120300: 'snowy-2',
        121000: 'rainy-2',
        121100: 'snowy-1',
        122000: 'rainy-3',
        122200: 'snowy-3',
        123000: 'rainy-1',
        123001: 'thunder',
        123002: 'rainy-1',
        123300: 'snowy-2',
        200000: 'night',
        210000: 'cloudy-night-1',
        210100: 'snowy-4',
        210300: 'snowy-5',
        211000: 'rainy-4',
        211100: 'snowy-5',
        213000: 'rainy-5',
        213300: 'snowy-4',
        220000: 'cloudy-night-1',
        220100: 'snowy-4',
        220200: 'snowy-6',
        220300: 'snowy-5',
        221000: 'rainy-4',
        221100: 'snowy-5',
        222000: 'rainy-6',
        222200: 'rainy-7',
        223000: 'rainy-5',
        223001: 'thunder',
        223002: 'rainy-5',
        223300: 'snowy-4',
        320000: 'cloudy',
        320100: 'snowy-4',
        320200: 'snowy-6',
        320300: 'snowy-5',
        321000: 'rainy-4',
        321100: 'snowy-5',
        322000: 'rainy-6',
        322200: 'rainy-7',
        323000: 'rainy-5',
        323001: 'thunder',
        323002: 'rainy-5',
        323300: 'snowy-4',
        330000: 'cloudy',
        330100: 'snowy-4',
        330200: 'snowy-6',
        330300: 'snowy-5',
        331000: 'rainy-4',
        331100: 'snowy-5',
        332000: 'rainy-6',
        332200: 'rainy-7',
        333000: 'rainy-5',
        333001: 'thunder',
        333300: 'snowy-4'
    },

    /**
     * getIconPath - Use the icon mapping to find the correct Amchart icon related to the given RBB
     * icon name. Returns the whole path to the SVG icon file based on the given subfolder. If no
     * mapping is found, a warning is logged and undefined will be returned.
     *
     * @param  {String} rbbIcon RBB icon name
     * @param  {String} folder  Amchart subfolder (animated or static)
     * @return {String}         Path to SVG icon
     */
    getIconPath: function(rbbIcon, folder) {

        const icon = this.iconMapping[rbbIcon];
        if (icon !== undefined) {
            return `vendor/amcharts/${folder}/${icon}.svg`;
        }

        this._log(`No mapping found for RBB icon "${rbbIcon}"!`);
        return undefined;
    },

    /**
     * _log - This logging wrapper is used so that it can be mocked in testing.
     *
     * @param  {String} message Message to log
     */
    _log: /* istanbul ignore next */ function(message) {
        console.warn(message);
    }
};


MMM-RBB-Weather.js
/* Magic Mirror
 * Module: MMM-RBB-Weather
 *
 * By Nikolai Keist (github.com/nkl-kst)
 * MIT Licensed.
 */

Module.register('MMM-RBB-Weather', {

    defaults: {

        // Data
        id: '10381', // Berlin-Steglitz
        days: 4, // Including current, 0 = only current data

        // Times
        animationSpeed: 1, // 1 sec.
        updateInterval: 600, // 10 min.

        // Show / hide flags
        showCurrentText: true,
        showCurrentWindspeed: true,
        showRainProbability: true,
        showUpdateTime: false,
        showWindspeed: false,

        // Styling
        animateCurrentIcon: true,
        animateForecastIcon: false,
        dayFormat: 'ddd',
        splitCurrentTextGreater: 30,
        tableClass: 'small',
        whiteIcons: true,

        // Trigger other modules
        triggers: [
            // eg. { day: 1, field: 'prr', value: 50, module: 'clock', hide: true }
        ]
    },

    // Requires Nunjucks support added in Magic Mirror 2.2.0
    requiresVersion: '2.2.0',

    // Instancevariable
    weatherData: null,
    updatedAt: null,

    getScripts: function() {
        return [
            'moment.js',
            'IconMapper.js'
        ];
    },

    getStyles: function() {
        return [
            'font-awesome.css',
            'font-awesome5.css', // Only available in MagicMirror > 2.5.0
            'weather-icons.css',
            'weather-icons-wind.css',
            'MMM-RBB-Weather.css'
        ];
    },

    getTranslations: function() {
        return {
            de: 'translations/de.json',
            en: 'translations/en.json'
        };
    },

    start: function() {
        Log.info(`Starting module: ${this.name} ...`);

        // Set locale
        moment.locale(this.config.language);
    },

    notificationReceived: function(notification, payload, sender) {

        // DOM ready
        if (notification === 'MODULE_DOM_CREATED') {

            // Initial data load
            this.loadData();
        }
    },

    socketNotificationReceived: function(notification, payload) {

        // Data loaded with node helper
        if (notification === 'DATA_LOADED') {
            this.weatherData = payload.data;
            this.updatedAt = payload.time;

            // Update module
            this.updateDom(this.config.animationSpeed * 1000);

            // Trigger other modules
            this.triggerModules();
        }
    },

    /**
     * getTemplate - Return the Nunjucks template that should be rendered. If there is no weather
     * data available yet, the 'nodata' template is returned.
     *
     * @return {String} Template to render
     */
    getTemplate: function() {

        // No data available
        if (this.weatherData === null || this.weatherData.length === 0) {
            return 'templates/nodata.njk';
        }

        // Default module template
        return 'templates/module.njk';
    },

    /**
     * getTemplateData - Return the data that is included in the rendered Nunjucks remplate. The
     * whole module instance is returned here, because several functions are called in the template.
     *
     * @return {Object} Data to put into templates
     */
    getTemplateData: function() {
        return { module: this };
    },

    /**
     * getCurrentText - Return the formatted current weather text. This is
     * split with a line break if the text contains a colon and length is
     * greater then config.splitCurrentTextGreater.
     *
     * @param  {String} text Current weather text
     * @return {String}      Formatted text
     */
    getCurrentText: function(text) {
        const splitValue = this.config.splitCurrentTextGreater;

        // No text available
        if (!text) {
            return undefined;
        }

        // Check if text and flag are given
        if (splitValue === 0) {
            return text;
        }

        // Check if text is long enough to split
        if (text.length <= splitValue) {
            return text;
        }

        // Split and break at colons
        return text.replace(/,\s*/, '<br/>');
    },

    /**
     * getDayText - Return the formatted day text for the given day index. This is used for forecast
     * days, so index 1 represents today, 2 represents tomorrow and so on.
     *
     * @param  {Number} dayIndex Day index
     * @return {String}          Formatted day text
     */
    getForecastDayText: function(dayIndex) {

        const day = moment().add(dayIndex - 1, 'days');
        return day.format(this.config.dayFormat);
    },

    /**
     * getIconUrl - Return URL to the animated or static icon mapped to the given RBB icon.
     *
     * @param  {Boolean} animate Use animated icons
     * @param  {String}  rbbIcon RBB icon to be mapped
     * @return {String}          URL to mapped icon
     */
    getIconUrl: function(animate, rbbIcon) {

        // No current icon available
        if (!rbbIcon || isNaN(rbbIcon)) {
            return undefined;
        }

        // Icon path
        const iconFolder = animate ? 'animated' : 'static';
        const iconPath = IconMapper.getIconPath(rbbIcon, iconFolder);

        // Fallback to RBB icons if no mapping was found
        if (!iconPath) {
            return `https://www.rbb24.de/basis/grafik/icons/wetter/svg/${rbbIcon}.svg`;
        }

        // Return icon url
        return this.file(iconPath);
    },

    /**
     * getTempIcon - Get thermometer icon depending on temperature value.
     *
     * @param  {Number} temp Temperature value
     * @return {String}      Icon name
     */
    getTempIcon: function(temp) {

        if (temp >= 35) return 'fa-umbrella-beach';
        if (temp >= 28) return 'fa-thermometer-full';
        if (temp >= 21) return 'fa-thermometer-three-quarters';
        if (temp >= 14) return 'fa-thermometer-half';
        if (temp >= 7) return 'fa-thermometer-quarter';
        if (temp >= 0) return 'fa-thermometer-empty';

        // Font Awesome 5 (with fa-snowflake) is not available in MagicMirror < 2.6.0
        if (version.localeCompare('2.6.0', 'en', { numeric: true }) < 0) {
            return 'fa-asterisk';
        }

        return 'fa-snowflake';
    },

    /**
     * getRainProbabilityIcon - Get rain icon depending on rain probability. Standard icon is
     * "fa-tint", this is dimmed if rain probability is <= 15% and the "fa-umbrella" icon is used
     * when probability is >= 70%.
     *
     * @param  {Number} prob Rain probability
     * @return {String}      Icon name
     */
    getRainProbabilityIcon: function(prob) {

        if (prob <= 15) {

            // Font Awesome 5 (with fa-tint-slash) is not available in MagicMirror < 2.6.0
            if (version.localeCompare('2.6.0', 'en', { numeric: true }) < 0) {
                return 'fa-tint dimmed';
            }

            return 'fa-tint-slash';
        }

        if (prob >= 70) {
            return 'fa-umbrella';
        }

        return 'fa-tint';
    },

    /**
     * getWindDirKey - Get wind direction short key based on wind direction degrees.
     *
     * @param  {String} deg Wind direction in degrees
     * @return {String}     Wind direction short text
     */
    getWindDirKey: function(deg) {

        // No wind direction available
        if (!deg || isNaN(deg)) {
            return undefined;
        }
        if (deg <= 22) return 'N';
        if (deg <= 67) return 'NE';
        if (deg <= 112) return 'E';
        if (deg <= 157) return 'SE';
        if (deg <= 202) return 'S';
        if (deg <= 247) return 'SW';
        if (deg <= 292) return 'W';
        if (deg <= 337) return 'NW';
        return 'N';
    },

    /**
     * getFormattedUpdatedTime - Get the human readable update time.
     *
     * @return {String} Formatted update time
     */
    getFormattedUpdateTime: function() {
        return moment(this.updatedAt).format('DD.MM. HH:mm');
    },

    /**
     * loadData - Load weather data via node_helper. This functions sends a socket notification with
     * LOAD_DATA as notification name and the place id (config.id) and day count (config.days) as
     * payload and schedules the next refresh.
     */
    loadData: function() {
        Log.info('Send socket notification to load data in node_helper ...');

        // Load data via node helper
        const dataConfig = { id: this.config.id, days: this.config.days };
        this.sendSocketNotification('LOAD_DATA', dataConfig);

        // Schedule next refresh
        this.scheduleRefresh();
    },

    /**
     * scheduleRefresh - Schedules refresh timer depending on config.updateInterval. This function
     * calls loadData().
     */
    scheduleRefresh: function() {

        setTimeout(() => {
            this.loadData();
        }, this.config.updateInterval * 1000);
    },

    /**
     * triggerModules - Trigger other modules based on weather data values.
     * Hides the defined module(s) if the data field is lower or equal than the
     * defined trigger value, otherwise shows the module. This behaviour is
     * switched if the hide-Property is used in triggers.
     */
    triggerModules: function() {

        // Do nothing if no data is available
        if (this.weatherData === null) return;

        // Iterate over defined triggers
        for (const trigger of this.config.triggers) {

            // Check if weather data is available for needed day
            const data = this.weatherData[trigger.day];
            if (!data) continue;

            // Get module(s) by class
            const modules = MM.getModules().withClass(trigger.module);
            for (const modul of modules) {
                Log.info(`Trigger module '${modul.name}' ...`);

                // Show/hide options
                const lockOptions = { lockString: this.name };
                const speed = this.config.animationSpeed * 1000;

                // Get weather data to check
                let dataValue = data[trigger.field];
                if (trigger.field === 'maxtemp') {
                    dataValue = data.temp.split(';')[0];
                }
                if (trigger.field === 'mintemp') {
                    dataValue = data.temp.split(';')[1];
                }

                // Check trigger condition: Hide module if data value is lower
                // or equal than trigger value or if the trigger uses hide mode,
                // but not both (xor)
                if ((dataValue <= trigger.value) !== (trigger.hide === true)) {
                    modul.hide(speed, lockOptions);
                } else {
                    modul.show(speed, lockOptions);
                }
            }
        }
    }
});
